File: Validation/py/CrossValidation.py
import numpy as np
def load_data(file):
	try:
		matrix = []
		f = open(file,'r')
		for line in f.readlines():
			x = list(map(float,line.split()))
			x.insert(0,1.0)
			matrix.append(x)
	finally:
		if f:
			f.close()
	return matrix

def sign(n):
	if n <=0:
		return -1
	else:
		return 1

def calErrRate(output,predict):
	errCount = 0
	for i in range(len(output)):
		if output[i]!=predict[i]:
			errCount = errCount + 1
	return errCount/len(output)

def minError(data):
	minError = 1
	for i in data:
		if i[1] < minError:
			minError = i[1]
			coef = i[0]
	return coef,minError

def getTrainAndValidationData(data,v,sequenceid):
	size = int(len(data)/v)
	print('split size:',size)
	traindata = []
	validationdata = []
	for i in range(len(data)):
		if i>=sequenceid*size and i<(sequenceid+1)*size:
			validationdata.append(data[i])
		else:
			traindata.append(data[i])
	return traindata,validationdata

def getDataXY(train,validation):
	colNum = len(train[0])
	trainx =[]
	trainy=[]
	for i in train:
		trainx.append(i[:colNum-1])
		trainy.append(i[colNum-1])
	validationx = []
	validationy = []
	for i in validation:
		validationx.append(i[:colNum-1])
		validationy.append(i[colNum-1])
	return trainx,trainy,validationx,validationy

def getAverageCrossValidationErrRate(data,coef,n):
	sumErrRate = 0
	colNum = len(data[0])
	for i in range(n):
		traindata,validationdata = getTrainAndValidationData(data,n,i)
		trainx,trainy,validationx,validationy = getDataXY(traindata,validationdata)
		weight = np.dot(np.dot(np.linalg.inv(np.add(np.dot(np.transpose(trainx),trainx),np.dot(coef,np.identity(colNum-1)))),np.transpose(trainx)),trainy)
		errRate = calErrRate(validationy,list(map(sign,np.dot(weight,np.transpose(validationx)))))
		sumErrRate = sumErrRate + errRate
	return sumErrRate/n

coef = [100,10,1,0.1,0.01,0.001,0.0001,0.00001,0.000001,0.0000001,0.00000001,0.000000001,0.0000000001]
cv = 5
data = load_data('train.txt')
train,validation = getTrainAndValidationData(data,3,0)
dataVal= []
for p in coef:
	errRate = getAverageCrossValidationErrRate(data,p,cv)
	coefIn = []
	coefIn.append(p)
	coefIn.append(errRate)
	dataVal.append(coefIn)
coef,errRate = minError(dataVal)
print('the coef:',coef)
print('the minimum',cv,'fold cross validation error rate:',errRate)

File: Validation/java/CrossValidation/src/CrossValidation.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.ejml.data.DenseMatrix64F;
import org.ejml.simple.SimpleMatrix;


public class CrossValidation {
	public static void main(String[] args){
		double minErrRate = 1;
		double coef =0;
		SimpleMatrix dataSimpleMatrix = loadData("train.txt");
		for (double i=2;i>=-10;i--){
			double errRate = getAverageValidationErrRate(dataSimpleMatrix,5,Math.pow(10, i));
			if (errRate < minErrRate){
				minErrRate = errRate;
				coef = Math.pow(10, i);
			}
		}
		System.out.println("the coef:"+coef);
		System.out.println("the minimum cross validation error rate:"+minErrRate);
	}
	public static double getAverageValidationErrRate(SimpleMatrix data,int cv,double coef){
		int rowNum = data.numRows();
		int colNum = data.numCols();
		int size = rowNum/cv;
		double sumErrRate = 0;
		for (int i=0;i<cv;i++){
			SimpleMatrix validationdata = null;
			SimpleMatrix traindata = null;
			validationdata = data.extractMatrix(size*i, size*(i+1), 0, colNum);
			if (i==0)
				traindata = data.extractMatrix(size*1, rowNum, 0, colNum);
			else
				traindata = data.extractMatrix(0, size*i, 0, colNum).combine(size*(i+1), 0, data.extractMatrix(size*(i+1), rowNum, 0, colNum));
			int trainRowNum = traindata.numRows();
			int trainColNum = traindata.numCols();
			SimpleMatrix trainx = traindata.extractMatrix(0, trainRowNum, 0, trainColNum-1);
			SimpleMatrix trainy = traindata.extractVector(false, trainColNum-1);
			//System.out.println(trainx);
			//System.out.println(trainy);
			SimpleMatrix weight = trainx.transpose().mult(trainx).plus(SimpleMatrix.identity(trainColNum-1).scale(coef)).invert().mult(trainx.transpose()).mult(trainy);
			double errRate = calculateErrorRate(weight,validationdata);
			sumErrRate =sumErrRate + errRate;
		}
		return (double) sumErrRate/cv;
	}
	public static double sign(double n){
		if (n<=0)
			return -1;
		else
			return 1;
	}
	public static double calculateErrorRate(SimpleMatrix w,SimpleMatrix data){
		int rowNum = data.numRows();
		int colNum = data.numCols();
		SimpleMatrix xdata = data.extractMatrix(0, rowNum, 0, colNum-1);
		SimpleMatrix ydata = data.extractVector(false, colNum-1);
		int errCount = 0;
		for (int i=0;i<rowNum;i++){
			double predict = sign(xdata.extractVector(true, i).dot(w.transpose()));
			if (predict!=ydata.get(i))
				errCount++;
;		}
		return (double) errCount/rowNum;
	}
	public static SimpleMatrix loadData(String fileName){
		List<String> dataList = new ArrayList<String>();
		BufferedReader br = null;
		try {
				br = new BufferedReader(new FileReader(fileName));
				String line = null;
				while((line = br.readLine())!=null){
					dataList.add(line.trim());
				}
				br.close();
		} catch (IOException e){
			e.printStackTrace();
		}
		int rowNum = dataList.size();
		int colNum = dataList.get(0).split("\\s+").length+1;
		double[] dataMatrix = new double[rowNum*colNum];
		for (int i=0;i<rowNum;i++){
			dataMatrix[i*colNum] = 1;
			String[] newLine = dataList.get(i).split("\\s+");
			//System.out.println(dataList.get(i));
			for (int j=0;j<newLine.length;j++){
				dataMatrix[i*colNum+j+1] = Double.valueOf(newLine[j]);
			}
		}
		DenseMatrix64F dataDenseMatrix = new DenseMatrix64F(rowNum,colNum,true,dataMatrix);
		SimpleMatrix dataSimpleMatrix = SimpleMatrix.wrap(dataDenseMatrix);
		return dataSimpleMatrix;
	}
}
